main/src/main/scala/sbt/EvaluateTask.scala

```diff
@@ -383,11 +383,13 @@ object EvaluateTask {
     streams(ScopedKey(Project.fillTaskAxis(key).scope, Keys.streams.key))
 
   def withStreams[T](structure: BuildStructure, state: State)(f: Streams => T): T = {
-    val str = std.Streams.closeable(structure.streams(state))
+    val logExchange = LogExchange
+    val str = std.Streams.closeable(structure.streams(state.put(logExchangeAttribute, logExchange)))
     try {
       f(str)
     } finally {
       str.close()
+      logExchange.close()
```

Comment:
I get that stream should be closed but I am not sure tying task evaluation, which is a part of act command, with `LogExchange` makes sense. Aren't there loggers that run the whole time?

Reply:
First off, I just noticed that I accidentally pushed a bug due to an incomplete cherry pick. This should be:
```patch
+    val logExchange = new LogExchange
+    val str = std.Streams.closeable(structure.streams(state.put(logExchangeAttribute, logExchange)))
```

> Aren't there loggers that run the whole time?

That's a great question. When I was trying to figure out how to safely remove appenders, I updated `LogExchange.logger` to print out the name of the logger it was creating. I then ran compile in https://github.com/jtjeferreira/sbt-multi-module-sample and found that it created about 6000 loggers for each run of compile. All of them had the format `streams-$i`. This made me believe that, in practice, what we needed to do was to clear out the loggers used during task evaluation. 

In my first attempt, I actually cached and removed all loggers from the `LogExchange` object but doing that to a global `LogExchange` seemed like a bad idea. That being said, in practice, it seemed to work fine but I didn't test it extensively. Instead, I just tried to figure out the boundaries where task evaluation began and task evaluation ended. My thinking is that it would be safe to drop all of the loggers created during task evaluation once the task evaluation completes. Part of what I was worried about was dynamic tasks. I was afraid that if a dynamic task completed it would clear _all_ the loggers not just the logger's it created. I convinced myself that this approach would work but it is a bit tricky to reason about. Another option I considered was to put similar logic in MainLoop rather than EvaluateTask.

My key assumption is that loggers are essentially not created outside of `GlobalLogging` and `Streams`. Users could theoretically be creating appenders outside of `streams.value.log` probably by using the `logManager` setting. My intuition is that that isn't really done. Even if it was, this change would only cause problems if they are caching the logger outside of the task evaluation and then try to use the cached logger outside of the task evaluation.

The `LogExchange` singleton is not cleared out between runs. GlobalLogging is initialized using the `LogExchange` singleton so it persists indefinitely for the life of the sbt process.

Reply:
I guess as a general direction, I am not comfortable with the idea of having multiple `LogExchange`s, since IMO it should be a thin wrapper around log4j. If we could keep the knowledge of stream loggers within the Stream that would be ideal, but if not maybe we could use State transition as a boundary? `MainLoop.scala` for example creates a new logger each time it processes before processing new command. If cleanup of arbitrarily created loggers are needed, I feel like that's the place to do it.

Reply:
I took a second pass at this and I'm convinced that a global `LogExchange` cannot work without leaking memory. The log4j `LoggerContext` data structure is append only (I verified this in the log4j source code) so even if we clean up the loggers as I've done in this PR, we will still have a memory leak in the `LoggerContext`. Due to this behavior, these changes are more of a bandaid than a fix. They reduce the amount that we leak but don't stop us from leaking.

The only way that I can see to prevent the resource leak is to create a new `LoggerContext` before every command evaluation and allow that context to go out of scope once command evaluation completes. I think your intuition is right that the place to manage the `LoggerContext` is in MainLoop. Prior to command evaluation, we'd add the new `LoggerContext` to the state via an attribute. When the command evaluation finishes, we'd remove the attribute so that the `LoggerContext` goes out of scope. I have this about 70% implemented, but I wanted to comment here first because I'll probably open a new PR since the implementation will be very different from this one.

Reply:
The tricky part might be allowing the `log.info(...)` to write into the appenders asynchronously, but wait for all appenders to finish writing before we close transient appenders out.

Reply:
How committed are we to using log4j? I am having a lot of problems getting rid of this leak because it seems nearly impossible to avoid creating global state internal to log4j. This ends up being a major problem for global logging because I cant' find anyway to clear out the ephemeral loggers without also blowing away global logging. I tried created different log4j configurations for both but somewhere underneath the hood they seem to end up sharing state. The result is that I cannot eliminate the memory leak without breaking global logging. I'm not going so far as to say that it's impossible to get working with log4j but I suspect that I can implement the subset of log4j that sbt needs faster than I can fix this bug.

Am I understanding the logging system correctly that the main reason for log4j is that the appender feature allows calls to, e.g. `log.info(msg)`, to be non-blocking and to fan out to multiple output sources? If so, that seems fairly straightforward to implement. If I were to go down this route, I'd refactor things in such a way that log4j could still be used as an optional implementation. It looks like log4j is only in the public api via `Keys.extraLoggers`.

Reply:
> How committed are we to using log4j?

I am not sure. In sbt 0.13 days, we had our own logging and I think it was performance bottleneck. One of the ideas we used to discuss internally was using a ring buffer to write logs in quickly, and let IO thread pick it up, based on Martin Thompson's talk. I chose log4j 2 because it adopted Martin Thompson's ring buffer library Disruptor, and at that time it was one of the fastest logger. At least it had the potential to be fast if configured correctly, which I may or may not have. Another benefit of sbt using log4j was that now plugins can use slf4j, and they would know that there's a binding for it.

If task logs are not good fit with log4j, I am open to changing it to something else.